### PR TITLE
Example Asset-based YAML DSL

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

This is an example demonstrating how to build a YAML DSL for asset graph construction

### CODE REVIEW

File: examples/experimental/assets_yaml_dsl/asset_yaml_dsl_tests/test_yaml_dsl.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    assets_def = assets_defs[0]
+    assert assets_def.key == AssetKey("asset_one")
+    assert assets_def.descriptions_by_key[assets_def.key] == "asset one description"
+
+
+def test_execution() -> None:
+    assets_defs = assets_defs_from_yaml("""
+assets:
+    - asset_key: asset_one
+      sql: "SELECT * from asset_one"
```
Comment (slopp) [inline examples/experimental/assets_yaml_dsl/asset_yaml_dsl_tests/test_yaml_dsl.py:66]: Suggestion: have this test include at least two assets, to avoid bugs of this flavor: https://github.com/dagster-io/dagster/discussions/11045#discussioncomment-4616243

**Review by slopp — APPROVED** [review]
I think this is a great opportunity for introing asset factories as a whole, with a specific YAML starting point. A nice formalization of https://github.com/dagster-io/dagster/discussions/11045.

(In fact, I'd go a far as to say approaching this as an asset factory example that happens to use YAML would be more valuable than a bespoke YAML DSL example).

A few thoughts:

- It would be great to flush out the README or create an "Assets Factory" guide. I think our examples often lack the prose to be approachable to newer users. But maybe thats a follow on to the RFC

- In terms of requirements, I think three things come to mind that aren't immediately obvious from this example:

(1) How would a non-technical user define these yaml specs and have dagster pick them up? This challenge is probably more bespoke to each org, but I think users often assume, want, or get confused about whether this YAML-loading can be done periodically, almost as a runtime thing, instead of as a build time step. They are used to a tool like Airflow "picking up" whatever files land in a dag directory in blob storage. In other words, if the person writing the YAML is totally separate from the Dagster team, what does the handoff look like? That'd be worth flushing out in a guide.

(2) Often the YAML-writer is interested in parameterizing a templated pipeline - or another way to think about it, often the YAML is at higher level than the asset graph, so one YAML chunk might represent a series of assets, not the 1:1 relationship shown here. How should users think about best representing these scenarios? Should the YAML translate to job-configuration for a pre-defined asset job? Or should the YAML "parameters" lead to the creation of a "copy" or "instantiation" of the templated assets? eg

(a) person A provides in YAML a list of stock tickers
(b) Each stock ticker should become an asset
(c) A final downstream asset should depend on all of (b)
(d) person B provides a different YAML list of stock tickers

(3) A follow up to 2, but perhaps the example could include a multi-asset as well? A common use case for asset factories is when some some job or sql snippet is run that results in an unknown # of assets or source assets being created.  It might be useful for this example to show how a multi-asset could be created for that case, and then how a downstream asset could depend on the dynamic upstreams



In short, the TL;DR is that this looks great, but might be worth considering:

What if the YAML users aren't writing YAML in the dagster repo? and What if their YAML is not a direct 1:1 with dagster assets?

I'd like @tacastillo 's perspective too, since he has long been an asset factory advocate.

**Comment by schrockn** [comment]
> It would be great to flush out the README or create an "Assets Factory" guide. I think our examples often lack the prose to be approachable to newer users. But maybe thats a follow on to the RFC

100%. That is where most of the value is. Wanted to get high-level feedback before investing heavily in docs.

> This challenge is probably more bespoke to each org, but I think users often assume, want, or get confused about whether this YAML-loading can be done periodically, almost as a runtime thing, instead of as a build time step.

This is a runtime thing. No build step required. Will make that clear in docs.

> Often the YAML-writer is interested in parameterizing a templated pipeline - or another way to think about it, often the YAML is at higher level than the asset graph, so one YAML chunk might represent a series of assets, not the 1:1 relationship shown here. How should users think about best representing these scenarios? 

Cool. I can add an example of what that would look like. Would be `multi_asset` which speaks to point 3.

> What if the YAML users aren't writing YAML in the dagster repo? 

Now this is an interesting question and related to what we were talking about yesterday @yuhan and @benpankow in terms of what support we would need to support a case where someone wants asset graphs defined in one repo to be reflected in another.

**Comment by slopp** [comment]
To clarify what I mean by "runtime" vs "build time"

IIUC in this example the YAML is loaded into definitions which are then deployed. I think some customers get confused about whether a new YAML requires a new deployment, or if they could set some of Dagster job to "read the YAMLs every hour". 

I think this pattern would require a re-deploy, which to me means it is a "build time" thing.

**Comment by schrockn** [comment]
Ah understood. Yes by that definition this is build time and I think is the pattern we should encourage and support.

**Comment by yuhan** [comment]
At a high level, this example looks pretty neat.

> What if the YAML users aren't writing YAML in the dagster repo?

This is a frequently asked question. Just spitballing - the protocol we're exploring could eventually be some format of "manifest" that gets checked in to the repo. From there, we could enable the cross-repo use case.

File: examples/experimental/assets_yaml_dsl/assets_yaml_dsl/stocks_dsl.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        # TODO: remove once alex's PR lands
+        return tuple([Output(None, ticker) for ticker in tickers])
+
+    @asset(deps=ticker_asset_keys, group_name=group_name)
+    def index_strategy() -> None:
+        stored_ticker_data = {}
+        for ticker in tickers:
+            stored_ticker_data[ticker] = fetch_data_for_ticker(ticker)
+
+        # do someting with stored_ticker_data
```
Comment (slopp) [inline examples/experimental/assets_yaml_dsl/assets_yaml_dsl/stocks_dsl.py:102]:
```suggestion
        # do something with stored_ticker_data
        if index_strategy == "weighted_average":
             index = sum([1/len(stored_ticker_data)*stored_ticker_data[i] fpr i in range(tickers)])
```

🤷

**Comment by slopp** [comment]
Really enjoying this example, I think it will open the eyes of many folks to showcase how flexible Dagster can be.

**Comment by schrockn** [comment]
Btw Lopp feel free to ignore while this is a work in progress (WIP).

**Comment by schrockn** [comment]
Current dependencies on/for this PR:
* master
  * **PR #15744** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15744" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/15744?utm_source=stack-comment).
